### ISSUE TITLE
af::arrray.device<>() returns NULL without eval

### ISSUE THREAD

**Issue by mricherzhagen** [arrayfire/arrayfire#1316]
Not sure if this is intended behavior. But if it is, it should be added to the documentation of `device()`:

```C++
af::setBackend(AF_BACKEND_CPU);

float v[] = {1, 2, 3, 4, 5};
array v_array = array(5, v);

array a_cpu = constant(0, 5,5);

a_cpu += matmulNT(v_array, v_array);
a_cpu.eval(); //Without this eval the device pointer is NULL. 
float *ptr = a_cpu.device<float>();
std::cout << "PTR: " << ptr << std::endl;
```

**Comment by FilipeMaia** [comment]
This looks like a bug. You should also check if you want to use `device()` or `af_get_raw_ptr()`.

**Comment by pavanky** [comment]
I am thinking this changed since 3.3.0 release. @mricherzhagen Can you run the CPU backend by setting environment variable `AF_SYNCHRONOUS_CALLS=1` ?

**Comment by shehzan10** [comment]
I just check this and getDevicePtr() in cpu/Array.hpp is not syncing before returning the pointer. CUDA/OpenCL sync/eval before returning the pointer.

**Comment by mricherzhagen** [comment]
I have to admit, i'm still on 3.2.2-1. Should have specified that earlier, sorry.

`AF_SYNCHRONOUS_CALLS=1` does not change anything for me.

**Comment by pavanky** [comment]
@mricherzhagen Can you check with 3.3.0, it may be *fixed*

**Comment by mricherzhagen** [comment]
I will try to do that, but i will probably not be able to until next week.

Another thing i just noticed: When i compile it without the call to `setBackend` and use `afopencl` it produces a segmentation fault without the `eval`. When the eval is present it doesn't. 
Same behavior with the unified backend, because it defaults to OpenCL before CPU, i guess. (I don't have CUDA at the moment)

**Comment by pavanky** [comment]
@mricherzhagen Does the OpenCL backend work otherwise?

**Comment by mricherzhagen** [comment]
I tried some of the examples and did not encounter any problems with the OpenCL backend

**Comment by mricherzhagen** [comment]
I compiled arrayfire from source using tag `v3.3.0` now.

I put this code into `examples/helloworld/devicePtr.cpp`:

```C++
#include<arrayfire.h>
#include<iostream>
using namespace af;

int main(int argc, char *argv[]) {
	info();
	float v[] = {1, 2, 3, 4, 5};
	array v_array = array(5, v);

	array a_cpu = constant(0, 5,5);

	a_cpu += matmulNT(v_array, v_array);
	if (argc > 1) 
		a_cpu.eval(); //Without this eval the device pointer is NULL. 
	float *ptr = a_cpu.device<float>();
	std::cout << "PTR: " << ptr << std::endl;
}
```

This is the result of running the program from the build directory:
```
$ examples/helloworld/devicePtr_cpu 
ArrayFire v3.3.0 (CPU, 64-bit Linux, build 5842ed2)
[0] Intel: Intel(R) Core(TM) i7 CPU       M 620  @ 2.67GHz, 7848 MB, Max threads(4) 
PTR: 0
$ examples/helloworld/devicePtr_cpu 1
ArrayFire v3.3.0 (CPU, 64-bit Linux, build 5842ed2)
[0] Intel: Intel(R) Core(TM) i7 CPU       M 620  @ 2.67GHz, 7848 MB, Max threads(4) 
PTR: 0x21368e0
```
As you can see: The problem still persists in v3.3.0

Apparently my OpenCl backend in my self-compiled version is kind of broken, but that is a different issue anyway, i think. (The array initialized from CPU memory was not copied to the GPU without eval?)

```
$ examples/helloworld/devicePtr_unified 
ArrayFire v3.2.2 (OpenCL, 64-bit Linux, build )
[0] NVIDIA  : NVS 3100M 
[1]    32380 segmentation fault (core dumped)  examples/helloworld/devicePtr_unified
$ examples/helloworld/devicePtr_unified 1
ArrayFire v3.2.2 (OpenCL, 64-bit Linux, build )
[0] NVIDIA  : NVS 3100M 
PTR: 0x24f2340
$ examples/helloworld/devicePtr_opencl   
examples/helloworld/devicePtr_opencl: /opt/cuda/lib64/libOpenCL.so.1: no version information available (required by ~/projects/arrayfire-git/build/src/backend/opencl/libafopencl.so.3)
examples/helloworld/devicePtr_opencl: /opt/cuda/lib64/libOpenCL.so.1: no version information available (required by ~/projects/arrayfire-git/build/src/backend/opencl/libafopencl.so.3)
ArrayFire v3.3.0 (OpenCL, 64-bit Linux, build 5842ed2)
[0] NVIDIA  : NVS 3100M, 511 MB
OpenCL error -11 on line 218 of ~/projects/arrayfire-git/build/third_party/clBLAS/src/clBLAS-ext/src/library/blas/xgemm.cc
clBuildProgram Failed
err = -11
Error: Failed to build program executable!

Build Log:

Error parsing value of -cl-std flag!
OpenCL error -45 on line 243 of ~/projects/arrayfire-git/build/third_party/clBLAS/src/clBLAS-ext/src/library/blas/xgemm.cc
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Unknown error:999):

In function af::array af::matmulNT(const af::array&, const af::array&)
In file src/api/cpp/blas.cpp:28
[1]    32431 abort (core dumped)  examples/helloworld/devicePtr_opencl
$ examples/helloworld/devicePtr_opencl 1
examples/helloworld/devicePtr_opencl: /opt/cuda/lib64/libOpenCL.so.1: no version information available (required by ~/projects/arrayfire-git/build/src/backend/opencl/libafopencl.so.3)
examples/helloworld/devicePtr_opencl: /opt/cuda/lib64/libOpenCL.so.1: no version information available (required by ~/projects/arrayfire-git/build/src/backend/opencl/libafopencl.so.3)
ArrayFire v3.3.0 (OpenCL, 64-bit Linux, build 5842ed2)
[0] NVIDIA  : NVS 3100M, 511 MB
OpenCL error -11 on line 218 of ~/projects/arrayfire-git/build/third_party/clBLAS/src/clBLAS-ext/src/library/blas/xgemm.cc
clBuildProgram Failed
err = -11
Error: Failed to build program executable!

Build Log:

Error parsing value of -cl-std flag!
OpenCL error -45 on line 243 of ~/projects/arrayfire-git/build/third_party/clBLAS/src/clBLAS-ext/src/library/blas/xgemm.cc
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Unknown error:999):

In function af::array af::matmulNT(const af::array&, const af::array&)
In file src/api/cpp/blas.cpp:28
[1]    32473 abort (core dumped)  examples/helloworld/devicePtr_opencl 1

```

**Comment by pavanky** [comment]
@mricherzhagen The OpenCL backend is failing because clBLAS (and arrayfire) requires OpenCL capability 1.2 or greater. The NVS 3100M only supports OpenCL 1.0.

As for the error in the CPU backend, this is going to be fixed by this PR:  https://github.com/arrayfire/arrayfire/pull/1322

**Comment by pavanky** [comment]
Fixed in hotfix 3.3.1 branch.
